sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/CreateTables.java

```diff
@@ -57,40 +61,71 @@
+      DynamicDestinations<?, DestinationT> dynamicDestinations) {
+    this(createDisposition, new BigQueryServicesImpl(), dynamicDestinations);
   }
 
   private CreateTables(
       CreateDisposition createDisposition,
       BigQueryServices bqServices,
-      SerializableFunction<TableDestination, TableSchema> schemaFunction) {
+      DynamicDestinations<?, DestinationT> dynamicDestinations) {
     this.createDisposition = createDisposition;
     this.bqServices = bqServices;
-    this.schemaFunction = schemaFunction;
+    this.dynamicDestinations = dynamicDestinations;
   }
 
-  CreateTables withTestServices(BigQueryServices bqServices) {
-    return new CreateTables(createDisposition, bqServices, schemaFunction);
+  CreateTables<DestinationT> withTestServices(BigQueryServices bqServices) {
+    return new CreateTables(createDisposition, bqServices, dynamicDestinations);
   }
 
   @Override
   public PCollection<KV<TableDestination, TableRow>> expand(
-      PCollection<KV<TableDestination, TableRow>> input) {
+      PCollection<KV<DestinationT, TableRow>> input) {
+    List<PCollectionView<?>> sideInputs = Lists.newArrayList();
+    if (dynamicDestinations.getSideInput() != null) {
+      sideInputs.add(dynamicDestinations.getSideInput());
+    }
     return input.apply(
         ParDo.of(
-            new DoFn<KV<TableDestination, TableRow>, KV<TableDestination, TableRow>>() {
-              @ProcessElement
-              public void processElement(ProcessContext context)
-                  throws InterruptedException, IOException {
-                BigQueryOptions options = context.getPipelineOptions().as(BigQueryOptions.class);
-                possibleCreateTable(options, context.element().getKey());
-                context.output(context.element());
-              }
-            }));
+                new DoFn<KV<DestinationT, TableRow>, KV<TableDestination, TableRow>>() {
+                  DynamicDestinations<?, DestinationT> dynamicDestinationsCopy;
+
+                  @StartBundle
+                  public void startBundle(Context c) {
+                    this.dynamicDestinationsCopy = SerializableUtils.clone(dynamicDestinations);
+                  }
+
+                  @ProcessElement
+                  public void processElement(ProcessContext context)
+                      throws InterruptedException, IOException {
+                    if (dynamicDestinationsCopy.getSideInput() != null) {
+                      dynamicDestinationsCopy.setSideInputValue(
+                          context.sideInput(dynamicDestinationsCopy.getSideInput()));
+                    }
+
+                    TableDestination tableDestination =
+                        dynamicDestinationsCopy.getTable(context.element().getKey());
+                    TableReference tableReference = tableDestination.getTableReference();
+                    if (Strings.isNullOrEmpty(tableReference.getProjectId())) {
```

Comment:
It is... it used to be in one place (in PrepareWrite), but we can no longer do it there since we don't have a TableDestination anymore.